src/Bicep.Core/Semantics/Namespaces/AzNamespaceType.cs

```diff
@@ -298,157 +299,188 @@ private static IEnumerable<FunctionOverload> GetAzOverloads(ResourceScope resour
-                .WithFlags(FunctionFlags.RequiresInlining)
-                .Build();
+            if (sourceFileKind == BicepSourceFileKind.ParamsFile)
+            {
+                yield return new FunctionOverloadBuilder("getSecret")
+                    .WithReturnType(LanguageConstants.String)
+                    .WithGenericDescription("Get Secret from KeyVault")
+                    .WithEvaluator(expression =>
+                    {
+                        var subscriptionIdParameter = (expression.Parameters[0] as StringLiteralExpression)?.Value;
+                        var resourceGroupNameParameter = (expression.Parameters[1] as StringLiteralExpression)?.Value;
+                        var keyVaultNameParameter = (expression.Parameters[2] as StringLiteralExpression)?.Value;
+                        var secretNameParameter = (expression.Parameters[3] as StringLiteralExpression)?.Value;
+                        var idExpression = ExpressionFactory.CreateObjectProperty("id", ExpressionFactory.CreateStringLiteral($"/subscriptions/{subscriptionIdParameter}/resourceGroups/{resourceGroupNameParameter}/providers/Microsoft.KeyVault/vaults/{keyVaultNameParameter}"));
+                        var keyVaultExpression = ExpressionFactory.CreateObjectProperty("keyVault", ExpressionFactory.CreateObject(new[] { idExpression }));
+                        var secretNameExpression = ExpressionFactory.CreateObjectProperty("secretName", ExpressionFactory.CreateStringLiteral($"{secretNameParameter}"));
+                        var referenceList = new List<ObjectPropertyExpression>() { keyVaultExpression, secretNameExpression };
+                        if (expression.Parameters.Length > 4)
+                        {
+                            referenceList.Add(ExpressionFactory.CreateObjectProperty("secretVersion", ExpressionFactory.CreateStringLiteral($"{(expression.Parameters[4] as StringLiteralExpression)?.Value}")));
+                        }
+                        return ExpressionFactory.CreateObject(new[] { ExpressionFactory.CreateObjectProperty("reference", ExpressionFactory.CreateObject(referenceList), expression.SourceSyntax) });
+                    })
+                    .WithRequiredParameter("subscriptionId", LanguageConstants.String, "Id of the Subscription that has the target KeyVault")
+                    .WithRequiredParameter("resourceGroupName", LanguageConstants.String, "Name of the Resource Group that has the target KeyVault")
+                    .WithRequiredParameter("keyVaultName", LanguageConstants.String, "Name of the target KeyVault")
+                    .WithRequiredParameter("secretName", LanguageConstants.String, "Name of the Secret")
+                    .WithOptionalParameter("secretVersion", LanguageConstants.String, "Version of the Secret")
+                    .Build();
```

Comment:
Can we add a validator here to make sure this function is only used as the right hand side of a `param` statement? Nested usages like `param invalid = '${az.getSecret(...)}'` should raise an error.

Reply:
Take a look at the usage of `FunctionFlags.ModuleSecureParameterOnly` to see how this is implemented for the existing `getSecret` function - I would want to try and align the two approaches.

Reply:
I just created a commit to add a new a FunctionFlag, and an ErrorDiagnostic. I also updated ValidatorVisitor 👍 `getSecret` function call is only valid if you directly assign it to a parameter now.